lib/livebook_web/live/app_session_live.ex

```diff
@@ -364,6 +365,16 @@ defmodule LivebookWeb.AppSessionLive do
     {:noreply, redirect_on_closed(socket)}
   end
 
+  def handle_info({:app_deployment_updated, %{slug: slug}}, %{assigns: %{slug: slug}} = socket) do
+    # We force the redirection in case of
+    # the current user loses access to this app.
+
+    # With this strategy, we guarantee that unauthorized users
+    # won't be able to keep reading the app which they
+    # should't have access.
+    {:noreply, redirect(socket, to: ~p"/apps/#{slug}/sessions/#{socket.assigns.session.id}")}
```

Comment:
This can cause the user to lose some state, like something they passed in the form. Could we redirect only if they lose access?